webgl2-support && tests changed

fabry24 opened the pull request:
Possible fix for #13529.

I have made a devices investigation and find out there are multiple newer devices not supporting OES_texture_float and many more not supporting OES_texture_float_linear, which makes NumpyTiles example and many more applications of this concept unuseful. However, many of these devices supports WebGL2, which natively support working with floating point by default. Also, there always could be devices not supporting OES_texture_float_linear, but we can fallback then to gl.NEAREST - as already implemented.

 Could you please make a revision of my merge request to get to main OL branch to support many more devices?

mike-000 commented on src/ol/webgl/TileTexture.js:
```diff
@@ -104,17 +110,19 @@ function uploadDataTexture(
     gl.TEXTURE_2D,
     0,
-    format,
+    internalFormat ? internalFormat : format,
```
Float32 should be supported with WebGL as well as WebGL2. Many PCs now running Windows 10 (upgraded from Windows 7) have no WebGL2 support but do support GL_OES_texture_float and GL_OES_texture_float_linear.

mike-000 commented on src/ol/webgl/TileTexture.js:
```diff
@@ -80,20 +80,26 @@ function uploadDataTexture(
       break;
     }
     case 3: {
+      internalFormat = data instanceof Float32Array ? RGB32F : gl.RGB;
```
Many PCs now running Windows 10 (upgraded from Windows 7) have no WebGL2 support but do support `GL_OES_texture_float` and `GL_OES_texture_float_linear`.

fabry24 replied:
Thank you, I have made a fix of it. Now it supports both webgl and webgl2.

tschaub commented:
@fabry24 - Thanks for your work on this.  The [rendering test failure](https://github.com/openlayers/openlayers/actions/runs/3639294257/jobs/6162723576) is from the [webgl-palette case](https://github.com/openlayers/openlayers/tree/main/test/rendering/cases/webgl-palette).

This is the expected output:

![expected](https://user-images.githubusercontent.com/41094/206549296-d59c4f14-ceb0-49dd-9591-55caf6649e04.png)

And here is the actual output in this branch:

![actual](https://user-images.githubusercontent.com/41094/206549394-245877c7-5cc4-41b5-8edc-bb4ce10e4fce.png)

You can run this single rendering test in "interactive" mode with the following:
```
node test/rendering/test.js --match webgl-palette --interactive
```

That will start the test server and use Chromium to render that case.  While the test server is running, you can open that same page with a different browser (http://127.0.0.1:3000/cases/webgl-palette/) for easier debugging.

mike-000 commented on src/ol/webgl/TileTexture.js:
```diff
@@ -80,20 +80,32 @@ function uploadDataTexture(
     case 1: {
       format = gl.LUMINANCE;
+      textureType = gl.UNSIGNED_BYTE;
+      data = new Uint8Array(Array.prototype.slice.call(data));
```
The failing test uses single band Float32 data but you are you are converting that to UInt8.  Since single band Float32 data is valid and can have values beyond the range of UInt8 that seems wrong.  Should the internal formats for 1 and 2 bands be R32F and RG32F for WebGL2 if the data is Float32?

tschaub commented:
I'm not that excited about the idea of trying to support both WebGL 1 and 2 if it means spreading fragile conditions across the library.

@fabry24 - Can you provide a bit more detail on what you have found regarding float texture and WebGL 2 support.  Ideally, a list of browsers/devices that we would gain.

I think it is worth considering a switch to WebGL 2 without providing a fallback.  It would be nice to have a list of what browsers/devices we would lose in this case.

The [OES_texture_float_linear](https://caniuse.com/mdn-api_oes_texture_float_linear) and [WebGL 2](https://caniuse.com/webgl2) coverage numbers make it look like we would drop from 97% to 94%.  But I'm not sure if that is the only way to look at it.

mike-000 commented:
Support is more related to hardware and drivers than to browsers.  On one desktop PC (Windows 7 upgraded to Windows 10) Chrome `chrome://gpu` and Edge `edge://gpu` report WebGL2 as unavailable and a WebGL2 context cannot be created.  On the same system Firefox shows WebGL2 in `about:support` and can use WebGL2.  But Chrome and Edge now dominate on Windows PCs and Firefox is rarely used.  As noted in https://github.com/openlayers/openlayers/issues/13500#issuecomment-1094976674 support for `OES_texture_float_linear` on older mobile devices is limited, although support on recent devices is good.  https://mikenunn.net/demo/testext.html will show a simple list of available extensions useful for screenshots via BrowserStack or LambdaTest.

fabry24 commented:
![20221209_211321](https://user-images.githubusercontent.com/78112061/206788643-a38ea8cb-44c4-4686-8b9a-09716c63a644.jpg)

We have found that there are mobile devices that support webgl2, which automatically includes support for oes_texture_float, but still do not support this library within webgl1. We tested the devices with the Chrome browser.
The list of non-functioning devices tested ( the numpytiles example from the openlayers examples did not work on these devices), but would work correctly within webgl2: 
Samsung A53,
Samsung Galaxy S21 Ultra,
Oppo Reno5 Z,
Xiaomi 11T,
Samsung Galaxy S10,
Samsung Galaxy S21,
Samsung Galaxy S21+

mike-000 commented:
> I'm not that excited about the idea of trying to support both WebGL 1 and 2 if it means spreading fragile conditions across the library.

It would simplify things a bit if LUMINANCE_ALPHA was never used with WebGL2

      internalFormat = webGL2 ? float ? gl.RG32F : gl.RG8 : gl.LUMINANCE_ALPHA;
      format = webGL2 ? gl.RG : gl.LUMINANCE_ALPHA;

Then we need to check for WebGL2 support anywhere else a `ol/has` property could be used.

fabry24 commented:
I have other devices that are not working properly
Samsung S8, S8+, S10e S20 and S21 Ultra

mike-000 commented:
Tests with Samsung devices can give variable results.  Samsung used either Mali or Adreno GPUs depending on the sales region, etc.  Only one of those has a problem with `OES_texture_float`.

fabry24 commented:
I tried to follow your instructions in supporting webgl2 to support also devices from other regions.
I have recognized problem with mocking HtmlCanvasElement.prototype.getContext, because of ol/has provides WEBGL2 support as constant - it is the same problem as DEVICE_PIXEL_RATIO constant. If I mock getContext function in test, there is already WEBGL2 constant evaluated as true - so actual state does not match ol/has constant. Should I make WEBGL2 constant as function?

mike-000 commented:
I had in mind something similar to the previous has `WEBGL` value https://github.com/openlayers/openlayers/commit/ef10834eb3816a7e489d18d711580debd4dbd0a6 which was not set until the first call was made.  Unfortunately the initial styles are parsed before the renderer is constructed, which would either mean constructing the renderer with an empty style then updating the style, which might be inefficient or, perhaps better, adding a new uniform for WebGL version so the `bandIndex` could be defined as

`bandIndex = '(u_version == 1) ? 3 : 1';`

(assuming that is valid GLSL syntax).

fabry24 commented:
in my last commit is fixed webgl2 support with fallback to webgl1 also with support of LUMINANCE and LUMINANCE_ALPHA. I have used Uniforms as you suggested. I added new tests for 6-band float32, also with HtmlCanvasElement.prototype.getContext() mocking test.
could you check it please?

mike-000 commented:
That looks good as it avoids breaking styles for UInt8 data.  However, 1 and 2 band float data with values in the 0 to 1 range such https://codesandbox.io/s/simple-forked-l6fyhn which will work with WebGL would still need explicit styling to work with WebGL2.  To avoid explicit styling a default color style could be set, which would also work for UInt8 data.  How to other reviewers feel about adding defaults?

On browsers which do not support WebGL2 `WebGL2RenderingContext` is undefined, so testing for it would throw an error.  A safer way to test for WebGL2 is when the context is not `WebGLRenderingContext`.

fabry24 commented:
I've changed the WebGL2RenderingContext check, also added default styles for single band and 2-bands inspired by your commit with test. Could I have anything more to do?

mike-000 commented:
The `RGB32F` and `RGBA32F` constants are not needed as they are also properties of the context similar to `gl.R32F` and `gl.RG32F`.

'internalFormat ? internalFormat : format,' could simply be `internalFormat,` as all cases now set a value for `internalFormat`.

An upgrade note might be needed.  With UInt8 data if the `bandCount` was left defaulted to 4 but only 1 or 2 bands provided in the data (as in https://github.com/openlayers/openlayers/blob/main/test/rendering/cases/webgl-data-tile-interpolate-false/main.js) it did and will still style correctly as 4 bands.  But with float data the correct `bandCount` will now need to be specified otherwise the output will differ depending on whether WebGL or WebGL2 is used.

mike-000 commented:
I think the original API description for `bandCount` remains sufficient.  But as a potential breaking change it should be mentioned in https://github.com/openlayers/openlayers/blob/main/changelog/upgrade-notes.md

The deleted comment originally at https://github.com/openlayers/openlayers/blob/main/src/ol/style/expressions.js#L507 is still relevant and would be helpful for anyone maintaining the code.

yurivlasov commented:
Do you have any idea in which version you can incorporate these changes? I agree with the discussion that it is a problem on many devices. Thank you

fabry24 commented:
Hello @mike-000 ,
Do you have any new news about this?
Are there any further adjustments needed?
Do you plan to include our changes in a release?
